Enhance volume-locality strategy

evanchaoli opened the pull request:
## Changes proposed by this PR

closes #8062

* [x] done

## Notes to reviewer

Enhanced performance of `volume-locality` strategy.

## Release Note

* Optimize database queries for `volume-locality` container placement strategy

evanchaoli committed:
enhance volume-locality strategy

Signed-off-by: Evan <chaol@vmware.com>

clarafu commented on atc/db/volume_repository.go:
```diff
@@ -465,6 +468,58 @@ func (repository *volumeRepository) FindResourceCacheVolume(workerName string, r
 }
 
+func (repository *volumeRepository) FindWorkersForResourceCache(resourceCache ResourceCache) ([]string, error) {
+	rows, err := psql.Select("distinct(c.worker_name)").
```
Hmm I think this will need some changes because it disregards the workers that are passed into the `Order(` method in the `placement.go` file. The list of workers that are passed into that method are a list of workers that are "compatible" (ex, non-team workers, tagged workers, etc) https://github.com/concourse/concourse/blob/dbbed7a7875441fe01784e05e485edb2079779c9/atc/worker/pool.go#L234-L237

evanchaoli replied:
Good cache.

clarafu replied:
Oh I was actually trying to describe that there is a problem that the list of workers that we query might not be a compatible worker and we were incrementing the `count[worker]++` in the map even if the worker is not a compatible worker. But I noticed you fixed it in your latest commit with https://github.com/concourse/concourse/blob/332618afd26014f612ecf199236e7201d9a16181/atc/worker/placement.go#L187 so great job!

evanchaoli replied:
@clarafu As mentioned in https://github.com/concourse/concourse/pull/8063#discussion_r806320698, incrementing the `count[worker]++` was actually not a bug, because sort is only done on candidate workers, and `count` map only provides supporting data to the sort.

clarafu replied:
Oh I'm sorry I didn't read that part properly earlier, thanks a lot for clarifying it to me again!

evanchaoli committed:
addressed review comments

Signed-off-by: Evan <chaol@vmware.com>

evanchaoli commented:
> Thx for the PR. Do you have the metrics comparison? Wondering the performance improvement over it.

Yes I have. In my load tests, initially I used "fewest-build-container", then I switched to "volume-locality". From then metrics, ATC CPU Usages were bumped to almost 100%, DB queries were also increase. When used "fewest-build-container", a build took around 1 minute to finish; after switched to "volume-locality", as ATC CPU usages were too high, builds got very slow, some builds took more than 1 hour. 

After I applied this PR and #8061 to the load test, ATC CPU usages dropped a lot, steadily below 50%. DB queries also decreased. And build duration reduced to less than 1 minute.

![8063](https://user-images.githubusercontent.com/18585861/154013161-f01af648-de4a-4e95-8d32-a351fd755831.png)

> Also could you add tests for those two new methods?

I think tests have been covered by `placement_test.go`. Looking at `volume_repository_test.go`, not all functions have tests in it. So I guess, we don't have duplicate tests in `volume_repository_test.go`.

To address @clarafu 's review comment, I found that current `volume-locality` strategy also has a bug. I fixed that, and added some tests in `placement_test.go`.

evanchaoli commented:
@xtremerui can you pls merge this one as it's approved?

xtremerui commented:
@clarafu is also reviewing.

clarafu reviewed (approved):
The improvements to the queries seem great!

The only thing I can think of that might cause a bit of inaccuracy with this change is the fact that sometimes there are worker resource caches or worker task caches that don't have a volume. Typically this should never happen but I'm guessing when something goes wrong on the worker side and the volume ends up getting deleted, we will end up with worker task caches or worker resource caches that have no volume associated to it. I saw this in a few workers in our deployments, but the workers I saw them on were usually ones that were special cased. But I don't think that should be a huge problem because worst case is that we choose a worker that doesn't actually have the volume on it, which is not that big of a deal. I would much rather have the performance improvements with this change.
